Windows: Callbacks - fixes bad callback validity check

dgmcdona opened the pull request:
This fixes a bug in the callbacks plugin which causes it to miss `IoRegisterShutdownNotification` callbacks on x86 samples. The `header.NameInfo.Name` field was being incorrectly treated as the device type. This fixes the issue by updating the `is_valid` method on the `_SHUTDOWN_PACKET` extension class to take a `type_map` parameter, and updates the method to correctly validate the object type.

LMK if we need to do a version bump on the callbacks plugin, and if so, what kind. I was unsure because the change to `is_valid`'s method signature is in the callbacks extensions module and not the plugin itself.

dgmcdona committed:
Windows: Fixes bad callback validity check

This fixes a bug in the callbacks plugin which causes it to miss
`IoRegisterShutdownNotification` callbacks on x86b samples. The
`header.NameInfo.Name` field was being incorrectly treated as the device
type. This fixes the issue by updating the `is_valid` method on the
`_SHUTDOWN_PACKET` extension type to take a `type_map` parameter, and
updates the method to correctly validate the object type.

atcuno commented:
@ikelos this is another one for the parity push

ikelos commented on volatility3/framework/symbols/windows/extensions/callbacks.py:
```diff
@@ -14,7 +15,7 @@ class _SHUTDOWN_PACKET(objects.StructType, pool.ExecutiveObject):
     """
 
-    def is_valid(self) -> bool:
+    def is_valid(self, type_map: Dict[int, str]) -> bool:
```
This is a change in API and will require bumps somewhere.  Please could you consider a different name, or determining the type_map within the call please?

dgmcdona replied:
Changed to `is_parseable(type_map)`

ikelos reviewed (commented):
Thanks, happy to get this fixed, but please don't go changing `is_valid`.  Changing it from every other `is_valid` method, and potentially for any other users of it is pretty messy and would require lots of version bumps.  Please implement a different method name, or determine the type_map from inside the call.

ikelos commented on volatility3/framework/symbols/windows/extensions/callbacks.py:
```diff
@@ -14,7 +15,7 @@ class _SHUTDOWN_PACKET(objects.StructType, pool.ExecutiveObject):
     It exposes a function which sanity-checks structure members.
     """
 
-    def is_valid(self) -> bool:
```
Did you get rid of `is_valid`?  That's also in the API, so would need a version bump to remove them?  I'd recommend `is_parsable` calls `is_valid` to replicate whatever it already did, and then just adds in the check based on the result at the end...

dgmcdona replied:
Done - I was torn on whether or not to duplicate the device object check in both methods and opted not to in order to avoid doing the pointer dereference twice, even if it makes the `is_valid` check somewhat less effective.

ikelos reviewed (commented):
Still needing a version bump in this iteration, I'd leave `is_valid` as it is (and make use of it in the new `is_parsable` function)...

dgmcdona committed:
Windows: Callbacks - fix breaking API change

Moves as much of the `is_parseable` check as possible back into an
`is_valid` method to avoid breaking API changes.

ikelos commented on volatility3/framework/symbols/windows/extensions/callbacks.py:
```diff
@@ -24,27 +25,54 @@ def is_valid(self) -> bool:
             return False
-
-        try:
-            header = device.get_object_header()
```
Looks like this test got remove from `is_valid`, which *could* return different results, but I'm kinda willing to overlook it since I imagine it would only affect a very small number of results.

ikelos reviewed (commented):
Thanks for making the requested changes.  I think it does still potentially alter the results that are returned, but I think it's likely a very small number of affected results.
